cml_data_tools/models.py

```diff
@@ -7,6 +9,99 @@
+    return components
+
+
+class AggregateIcaModel:
+    """Transforms a list of models and a list of clusters (cf. `iter_clusters`
+    in clustering.py) into a model capable of transforming using the exemplars
+    of each component cluster across the various submodels.
+
+    Parameters
+    ----------
+    models
+        A list of IcaPhenotypeModel instances in order
+    clusters
+        A list of 3-tuples, the third of which is the exemplar ordinal
+        I.e. if the cluster exemplar is the 300th phenotype of the 4th model
+        then the ordinal of that exemplar is (3 * n) + 299, where `n` is the
+        number of phenotypes that was requested across the submodels
+    n : Int
+        The number of phenotypes requested per submodel
+    """
+    def __init__(self, models, clusters, n=2000):
+        self.models = [self.ModelCore(m) for m in models]
+        self.components = clusters_to_components(clusters, n=n)
+        # Sanity check
+        assert len(self.models) == len(self.components)
+
+        # Get the Phenotypes of the exemplars and stitch them together
+        # We do this on init to avoid keeping a reference to models anywhere
+        parts = []
+        for i, (model, ids) in enumerate(zip(models, self.components)):
+            cols = np.array(model.phenotype_names)[ids]
+            ph = model.phenotypes_[cols]
+            ph.columns = self.ids_to_names(i, ids)
+            parts.append(ph)
+        self.phenotypes = pd.concat(parts, axis=1)
+
+    def ids_to_names(self, m, ids):
+        """
+        Produces an array of appropriate identifiers for the phenotypes
+        identified by model_id `m` and column indices `ids`
+        """
+        return np.array([f'M{m:03}-P{c:04}' for c in ids])
+
+    def transform(self, X):
+        """Project the data in X onto the aggregated phenotype cluster
+        exemplars.
+
+        Arguments
+        ---------
+        X : pandas.DataFrame
+            Has one row per instance, one column per variable
+
+        Returns
+        -------
+        pandas.DataFrame
+            A pandas dataframe with rows corresponding to (and indexed the same
+            as) rows in X, and columns corresponding to the aggregated
+            phenotypes. Since the exemplar names may not be unique across the
+            submodels, the columns are *renamed* from the original
+            Each cell contains the amount of the given phenotype expressed by
+            the row of X.
+        """
+        parts = []
+        for i, (model, ids) in enumerate(zip(self.models, self.components)):
+            columns = np.array(model.phenotype_names)[ids]
+            x_hat = model.transform(X)[columns]
+            x_hat.columns = self.ids_to_names(i, ids)
+            parts.append(x_hat)
+        return pd.concat(parts, axis=1)
+
+    class ModelCore:
+        """Extracts core functionality from IcaPhenotypeModel"""
+        def __init__(self, model):
+            self.ica = model.ica
```

Comment:
Do the reference semantics work here? Meaning, does `model` need to be kept around for this to work?

Reply:
Yes, this copies the attributes from `model` to the `ModelCore` instance, which allows us to save this separately without needing to keep the entire `model` instance around.